### ISSUE TITLE
Rule verification false positives

### ISSUE THREAD

**Issue by lukpueh** [in-toto/in-toto#261]
**Description of issue or feature request**:

The firewall like processing of artifact rules in `verify_item_rules` uses a material queue, a product queue and a generic artifact queue. If one of the sequentially processed rules can be applied successfully, the corresponding artifacts in the corresponding queue are removed.

The queues are updated in different places, e.g. `verify_match_rule` modifies the artifacts queue at the passed reference, which btw. also points to the materials or products queue in the beginning, only to override itself upon return. Other `verify_*_rule` function only return a new updated queue to override the old queue. Depending on the rule type ("match", "allow", ...) and the rule source type ("materials" or "products"), the three queues further update each other.

Besides being hard to understand the rule processing may lead to false positives, i.e. it may fail rule verification although it shouldn't. Here is one symptomatic minimal failing example:

```python
from in_toto.models.metadata import Metablock
from in_toto.models.link import Link
from in_toto import verifylib

links = {
  "item": Metablock(signed=Link(name="item",
       products={
           "foo" : {"sha256": "deadbeef"},
       }
  )),
}

rules = [
  # ALLOW does not consume anything here but cuts reference between
  # `source_artifacts_queue` and `source_products_queue`. NOTE: Without this
  # ALLOW the behavior of the rule sequence would be as expected.
  ["ALLOW", "bogus"],

  # MATCH consumes `foo` and removes it from `source_artifact_queues`
  ["MATCH", "foo", "WITH", "PRODUCTS", "FROM", "item"],

  # CREATE does not consume anything here, and overrides
  # `source_artifacts_queue` with an outdated `source_products_queue`
  # i.e. undoes MATCH from above.
  ["CREATE", "bogus"],

  # DISALLOW fails due to `foo` being in the `source_artifacts_queue` again
  # NOTE: CREATE re-establishes the reference between the two queues, thus
  # repeating the MATCH, CREATE rule sequence before the DISALLOW rule would
  # go back to expected behavior.
  ["DISALLOW", "*"]
]

verifylib.verify_item_rules("item", "products", rules, links)
```

**Current behavior**:
Rule processing implementation is unnecessarily complicated and may lead to false positives under certain circumstances.

**Expected behavior**:
Fix false positives and simplify rule processing.

**Comment by lukpueh** [comment]
Replacing “products” with “materials” (also in the match rule) and the `CREATE` rule with a `DELETE` rule above would result in a similar false positive. And it should be possible to craft another similar false positive example with the `MODIFY` rule, because it also relies on the materials and products queues.

**Comment by trishankatdatadog** [comment]
VERY good find, Lukas!

**Comment by trishankatdatadog** [comment]
Are false negatives possible? Worth taking a look...

**Comment by lukpueh** [comment]
AFAICS there is no scenario where a queue is accidentally modified to have fewer artifacts than it should, and the `DISALLOW` rule seems to work, hence there should be no false negatives.

However, my re-write of the rule processing functions in #262 should make it easier to rule out false positives and false negatives.

**Comment by lukpueh** [comment]
Fixed with #262
